File: client-scalaz/src/main/scala/org/openapitools/client/api/Pack.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime

import java.time.OffsetDateTime

import Pack._

case class Pack (
  id: Option[Long],
active: Option[Boolean],
valid: Option[Boolean],
newOwnerId: Option[Long],
owner: Option[Account],
application: Option[Application],
sequenceType: Option[SequenceType],
title: Option[String],
description: Option[String],
startDate: Option[OffsetDateTime],
endDate: Option[OffsetDateTime],
packType: Option[PackType],
image: Option[Asset],
background: Option[Asset],
imageURL: Option[String],
gameLevels: Option[List[GameLevel]],
hasMoreItems: Option[Boolean],
totalCount: Option[Long],
downloaded: Option[Boolean],
authorOverride: Option[String],
packOrder: Option[Long],
inGame: Option[Boolean],
highest: Option[Boolean],
notificationCount: Option[Integer],
points: Option[Long],
ticketType: Option[String],
ticketCount: Option[Long],
priceType: Option[String],
price: Option[Integer],
allocateTickets: Option[Boolean],
applicationTitle: Option[String],
levelNumberMap: Option[Map[String, GameLevel]])

object Pack {
  import DateTimeCodecs._
  sealed trait SequenceType
  case object FIRSTAVAILABLE extends SequenceType
  case object ALLAVAILABLE extends SequenceType

  object SequenceType {
    def toSequenceType(s: String): Option[SequenceType] = s match {
      case "FIRSTAVAILABLE" => Some(FIRSTAVAILABLE)
      case "ALLAVAILABLE" => Some(ALLAVAILABLE)
      case _ => None
    }

    def fromSequenceType(x: SequenceType): String = x match {
      case FIRSTAVAILABLE => "FIRSTAVAILABLE"
      case ALLAVAILABLE => "ALLAVAILABLE"
    }
  }

  implicit val SequenceTypeEnumEncoder: EncodeJson[SequenceType] =
    EncodeJson[SequenceType](is => StringEncodeJson(SequenceType.fromSequenceType(is)))

  implicit val SequenceTypeEnumDecoder: DecodeJson[SequenceType] =
    DecodeJson.optionDecoder[SequenceType](n => n.string.flatMap(jStr => SequenceType.toSequenceType(jStr)), "SequenceType failed to de-serialize")
  sealed trait PackType
  case object TUTORIAL extends PackType
  case object BUILTIN extends PackType
  case object DOWNLOAD extends PackType
  case object THRESHOLD extends PackType
  case object THEME extends PackType
  case object TOURNAMENT extends PackType

  object PackType {
    def toPackType(s: String): Option[PackType] = s match {
      case "TUTORIAL" => Some(TUTORIAL)
      case "BUILTIN" => Some(BUILTIN)
      case "DOWNLOAD" => Some(DOWNLOAD)
      case "THRESHOLD" => Some(THRESHOLD)
      case "THEME" => Some(THEME)
      case "TOURNAMENT" => Some(TOURNAMENT)
      case _ => None
    }

    def fromPackType(x: PackType): String = x match {
      case TUTORIAL => "TUTORIAL"
      case BUILTIN => "BUILTIN"
      case DOWNLOAD => "DOWNLOAD"
      case THRESHOLD => "THRESHOLD"
      case THEME => "THEME"
      case TOURNAMENT => "TOURNAMENT"
    }
  }

  implicit val PackTypeEnumEncoder: EncodeJson[PackType] =
    EncodeJson[PackType](is => StringEncodeJson(PackType.fromPackType(is)))

  implicit val PackTypeEnumDecoder: DecodeJson[PackType] =
    DecodeJson.optionDecoder[PackType](n => n.string.flatMap(jStr => PackType.toPackType(jStr)), "PackType failed to de-serialize")

  implicit val PackCodecJson: CodecJson[Pack] = CodecJson.derive[Pack]
  implicit val PackDecoder: EntityDecoder[Pack] = jsonOf[Pack]
  implicit val PackEncoder: EntityEncoder[Pack] = jsonEncoderOf[Pack]
}

File: client-clojure/src/sirqul_io_t_platform/specs/pack.clj
(ns sirqul-io-t-platform.specs.pack
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.account :refer :all]
            [sirqul-io-t-platform.specs.application :refer :all]
            [sirqul-io-t-platform.specs.asset :refer :all]
            [sirqul-io-t-platform.specs.asset :refer :all]
            [sirqul-io-t-platform.specs.game-level :refer :all]
            [sirqul-io-t-platform.specs.game-level :refer :all]
            )
  (:import (java.io File)))


(def pack-data
  {
   (ds/opt :id) int?
   (ds/opt :active) boolean?
   (ds/opt :valid) boolean?
   (ds/opt :newOwnerId) int?
   (ds/opt :owner) account-spec
   (ds/opt :application) application-spec
   (ds/opt :sequenceType) string?
   (ds/opt :title) string?
   (ds/opt :description) string?
   (ds/opt :startDate) inst?
   (ds/opt :endDate) inst?
   (ds/opt :packType) string?
   (ds/opt :image) asset-spec
   (ds/opt :background) asset-spec
   (ds/opt :imageURL) string?
   (ds/opt :gameLevels) (s/coll-of game-level-spec)
   (ds/opt :hasMoreItems) boolean?
   (ds/opt :totalCount) int?
   (ds/opt :downloaded) boolean?
   (ds/opt :authorOverride) string?
   (ds/opt :packOrder) int?
   (ds/opt :inGame) boolean?
   (ds/opt :highest) boolean?
   (ds/opt :notificationCount) int?
   (ds/opt :points) int?
   (ds/opt :ticketType) string?
   (ds/opt :ticketCount) int?
   (ds/opt :priceType) string?
   (ds/opt :price) int?
   (ds/opt :allocateTickets) boolean?
   (ds/opt :applicationTitle) string?
   (ds/opt :levelNumberMap) (s/map-of string? game-level-spec)
   })

(def pack-spec
  (ds/spec
    {:name ::pack
     :spec pack-data}))
